compiler/test/dotty/tools/scripting/BashExitCodeTests.scala

```diff
@@ -0,0 +1,156 @@
+   */
+  private def testCommandExitCode(args: Seq[String], expectedExitCode: Int): Unit =
+    val commandline = args.mkString(" ")
+    val (validTest, exitCode, output, erroutput) = bashCommand(commandline)
+    if verifyValid(validTest) then
+      assertEquals(expectedExitCode, exitCode)
+
+
+  /* Compiles and then runs the code checking the exit code
+   *
+   * @param testFileName name of the test
+   * @param runExitCode expected exit code from the runner output
+   */
+  private def testCompileAndRun(testFileName: String, runExitCode: Int): Unit =
+    val outputDirectory = testCompilationExitCode(testFileName, 0)
+
+    def testRuntimeExitCode(className: String, expectedExitCode: Int): Unit =
+        val testClassFile = outputDirectory.files.find(_.getName == s"$className.class")
+        assert(testClassFile.isDefined)
+        val commandline = (Seq(scalaPath, "-classpath", outputDirectory.getAbsolutePath, className)).mkString(" ")
+        val (validTest, exitCode, o, e) = bashCommand(commandline)
+          if verifyValid(validTest) then
+            assertEquals(expectedExitCode, exitCode)
+
+    try {
+      val generatedClassName = testFileName.split('.').head
+      testRuntimeExitCode(generatedClassName,  runExitCode)
+    } finally {
+      Util.deleteFile(outputDirectory)
+    }
+
+  /*
+   * Checks if scripting test resources contains test with given `testName`
+   * And then runs function `test`
+   *
+   * @param testName name of the test containing the extension
+   * @param test check to be run on found test file
+   */
+  private def assertTestExists(testName: String)(test: File => Unit) =
+    val file = testFiles.find(_.getName == testName)
+    assert(file.isDefined)
+    test(file.get)
+
+class BashExitCodeTests:
+  import BashExitCodeTests.*
+
+  @Test def verifyExitCodeOnCompileError: Unit =
+    testCompilationExitCode("compileError.scala", 1, true)
+
+  @Test def verifyExitCodeOnRuntimeError: Unit =
+    testCompileAndRun("runtimeError.scala", 1)
+
+  @Test def verifyExitCode: Unit =
+    testCompileAndRun("positiveTest.scala", 0)
+
+  @Test def verifyExitCodeOnScriptError: Unit =
+    assertTestExists("scriptRuntimeError.sc") { file =>
+      testCommandExitCode(Seq(scalacPath, "-script", file.absPath), 1)
+    }
+
+  @Test def verifyExitCodeOnScriptErrorCompiler: Unit =
+    assertTestExists("scriptRuntimeError.sc") { file =>
+      testCommandExitCode(Seq(scalacPath, "-script", file.absPath), 1)
+    }
+
+  @Test def verifyExitCodeOnScript: Unit =
+    assertTestExists("scriptPositive.sc") { file =>
+      testCommandExitCode(Seq(scalaPath, file.absPath), 0)
+    }
+
+  @Test def verifyExitCodeOnScriptCompiler: Unit =
+    assertTestExists("scriptPositive.sc") { file =>
+      testCommandExitCode(Seq(scalacPath, "-script", file.absPath), 0)
+    }
+
+  @Test def verifyExitCodeOnDecompilation: Unit =
+    val testName = "positiveTest"
+    val outputDirectory = testCompilationExitCode(s"$testName.scala", 0)
```

Comment:
this doesn't seem to be deleted - same with for `verifyExitCodeOnPrintTasty`

Reply:
I refactored it to make it simpler to create tests and always guarantee temporary file deletion